Week3/sql-injection.js

```diff
@@ -0,0 +1,50 @@
+var prompt = require('prompt');
+var mysql      = require('mysql');
+const util = require('util');
+
+const connection = mysql.createConnection({
+  host     : 'localhost',
+  user     : 'hyfuser',
+  password : 'hyfpassword',
+  database : 'class17',
+  multipleStatements: true
+});
+
+const execQuery = util.promisify(connection.query.bind(connection))
+const input = util.promisify(prompt.get.bind(this))
+
+async function queryDatabase() {
+
+    var input_number = ""
+    prompt.start();
+    try {
+        const result = await input(['number']);
+        input_number = result.number
+    } catch(err) {
+        console.error(err);
+    }
+
+    // 1. Naive way of passing the parameter to the query
+    //const select_query = `select * from students WHERE student_number =  ${input_number};`
+
+    // 2. Escaping the parameter ( replacing the unwanted characters)
+    //const select_query = `select * from students WHERE student_number =` + connection.escape(input_number);
+    
+    // 3. Using a question mark syntax to do the escaping (AKA prepared statements)
+    const select_query = `select * from students WHERE student_number = ?`
+
+    connection.connect();
+    try {
+        console.log(select_query);
+        var results = await execQuery(select_query, input_number);
+    } catch(error) {
+        console.error(error);
+    }
+    
+    for (i in results) {
+        console.log(results[i]);
+    }
+    connection.end();
+}
+
+queryDatabase();
```

Comment:
You have two try/catch blocks in a single function. There should only be one. Note also that in JS we teach that `for...in` should not be used with arrays (or preferably not used in any circumstance). We teach to use `for...of` or the `.forEach()` array method instead:

```js
async function queryDatabase() {
  connection.connect();

  try {
    prompt.start();
    const result = await input(['number']);
    const inputNumber = result.number;

    // 1. Naive way of passing the parameter to the query
    //const select_query = `select * from students WHERE student_number =  ${input_number};`

    // 2. Escaping the parameter ( replacing the unwanted characters)
    //const select_query = `select * from students WHERE student_number =` + connection.escape(input_number);

    // 3. Using a question mark syntax to do the escaping (AKA prepared statements)
    const select_query = `select * from students WHERE student_number = ?`;

    console.log(select_query);
    const results = await execQuery(select_query, inputNumber);

    for (const result of results) {
      console.log(results);
    }
  } catch (error) {
    console.error(error);
  }

  connection.end();
}
```